src/runtime/cuda/cuda_profiler.cpp

```diff
@@ -0,0 +1,156 @@
+/*
+ * This file is part of hipSYCL, a SYCL implementation based on CUDA/HIP
+ *
+ * Copyright (c) 2019-2020 Aksel Alpay and contributors
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "hipSYCL/runtime/cuda/cuda_profiler.hpp"
+#include "hipSYCL/runtime/error.hpp"
+#include "hipSYCL/runtime/cuda/cuda_event.hpp"
+
+#include <cuda_runtime_api.h>
+
+#include <cassert>
+#include <memory>
+
+namespace hipsycl {
+namespace rt {
+
+namespace {
+
+cuda_unique_event async_record_event(cudaStream_t stream) {
+  auto evt = make_cuda_event();
```

Comment:
Is it guaranteed that the device has been activated here?

Reply:
That should probably be a precondition of `async_record_event`. I can see that `cuda_queue::submit_kernel()` activates the device before transitively calling this function, while `submit_memcpy` does not, so activation is probably missing in that case. Does `cudaMemcpyAsync` / `cudaMemsetAsync` not need device activation?

Reply:
The device is now properly activated in all cases before the call to `async_record_event`.

Reply:
> I can see that cuda_queue::submit_kernel() activates the device before transitively calling this function, while submit_memcpy does not, so activation is probably missing in that case. Does cudaMemcpyAsync / cudaMemsetAsync not need device activation?

No, and I don't think it's required for kernel submission either (it may be an oversight that it's done there). The reason is that streams are already bound to a particular device, so if you submit to a stream on a given device it is not necessary to additionally set the device.

However, I've learned the hard way that event construction is already tied to a particular device, i.e. you need to have the right device activated when an event is constructed. This was only recently fixed in #399, and this is why I was asking.